demosplan/DemosPlanCoreBundle/Tools/PdfCreatorRabbitmq.php

```diff
@@ -0,0 +1,120 @@
+
+class PdfCreatorRabbitmq implements PdfCreatorInterface
+{
+    /**
+     * @var RpcClient
+     */
+    protected $client;
+
+    public function __construct(private GlobalConfigInterface $globalConfig, private LoggerInterface $logger)
+    {
+    }
+
+    /**
+     * Erstelle ein PDF aus einer tex-Vorlage.
+     *
+     * @param string $content  base64 encodierte tex-Datei
+     * @param array  $pictures array der Form ['picture0 => base64_encode(''), 'picture1' => ....]
+     *
+     * @throws Exception
+     */
+    public function createPdf(string $content, array $pictures = []): string
+    {
+        $payload = [
+            'file' => $content,
+        ];
+        $payload = array_merge($payload, $pictures);
+        $msg = Json::encode($payload);
+
+        $this->logger->debug(
+            'Export pdf with RabbitMQ, with routingKey: '.$this->globalConfig->getProjectPrefix()
+        );
+        $this->logger->debug(
+            'Content to send to RabbitMQ: '.DemosPlanTools::varExport(
+                base64_decode($content),
+                true
+            )
+        );
+        $this->logger->debug(
+            'Number of pictures send to RabbitMQ: '.count($pictures)
+        );
+
+        try {
+            $routingKey = $this->globalConfig->getProjectPrefix();
+            if ($this->globalConfig->isMessageQueueRoutingDisabled()) {
+                $routingKey = '';
+            }
+            $this->client->addRequest(
+                $msg,
+                'pdfDemosPlan',
+                'exportPDF',
+                $routingKey,
+                600
+            );
+            $replies = $this->client->getReplies();
+            $this->logger->debug(
+                'Got replies ',
+                [DemosPlanTools::varExport($replies, true)]
+            );
+
+            $exportResult = Json::decodeToArray($replies['exportPDF']);
+            if (null === $exportResult) {
+                $this->logger->error(
+                    'Reply from RabbitMQ: ',
+                    [DemosPlanTools::varExport($replies, true)]
+                );
+                throw new Exception('Could not decode export result');
+            } elseif (!isset($exportResult['file'])) {
+                $this->logger->error(
+                    'AMPQResult has wrong format ',
+                    [DemosPlanTools::varExport($exportResult, true)]
+                );
+                throw new Exception('AMPQResult has wrong format');
+            }
+
+            return $exportResult['file'];
+        } catch (AMQPTimeoutException $e) {
+            $this->logger->error(
+                'Fehler in ImportConsumer:',
+                [$e]
+            );
+            throw new TimeoutException('Timeout ');
+        } catch (Exception $e) {
+            $this->logger->error(
+                'Could not create PDF ',
+                [$e]
+            );
+            throw new Exception('Could not create PDF ');
```

Comment:
Just once here: I'd prefer having a PdfCreateException and static constructors for the different cases, but I am unsure whether this is part of our coding guidelines or just something I've seen and done a bunch of times.

Reply:
I am quite sure that this is currently not part of our coding guidelines as this has never been discussed (at least I am not aware of it and I dislike the static constructor). But having specialized Exceptions is better